src/eager/readonly_tensor.rs

```diff
@@ -0,0 +1,151 @@
+use libc::c_int;
+use std::ops::Deref;
+use tensorflow_sys as tf;
+
+use crate::{AnyTensor, DataType, Result, Shape, Tensor, TensorInner, TensorType};
+
+/// A read-only tensor.
+///
+/// ReadonlyTensor is a [`Tensor`](Tensor) that does not support mutation.
+#[derive(Debug, Clone, Eq)]
+pub struct ReadonlyTensor<T: TensorType> {
+    pub(super) inner: T::InnerType,
+    pub(super) dims: Vec<u64>,
+}
+
+impl<T: TensorType> AnyTensor for ReadonlyTensor<T> {
+    fn inner(&self) -> Result<*mut tf::TF_Tensor> {
+        self.inner.as_mut_ptr(&self.dims)
+    }
+
+    fn data_type(&self) -> DataType {
+        T::data_type()
+    }
+}
+
+impl<T: TensorType> Deref for ReadonlyTensor<T> {
+    type Target = [T];
+
+    #[inline]
+    fn deref(&self) -> &[T] {
+        self.inner.deref()
+    }
+}
+
+impl<T: TensorType + PartialEq> PartialEq for ReadonlyTensor<T> {
+    fn eq(&self, other: &Self) -> bool {
+        self.dims == other.dims && self.deref() == other.deref()
+    }
+}
+
+impl<T: TensorType + PartialEq> PartialEq<Tensor<T>> for ReadonlyTensor<T> {
+    fn eq(&self, other: &Tensor<T>) -> bool {
+        self.dims == other.dims && self.deref() == other.deref()
+    }
+}
+
+impl<T: TensorType> ReadonlyTensor<T> {
+    /// Get one single value from the Tensor.
+    ///
+    /// ```
+    /// # use tensorflow::Tensor;
+    /// # use tensorflow::eager::ReadonlyTensor;
+    /// let mut a = Tensor::<i32>::new(&[2, 3, 5]);
+    ///
+    /// a[1*15 + 1*5 + 1] = 5;
+    /// let a: ReadonlyTensor<_> = a.freeze();
+    /// assert_eq!(a.get(&[1, 1, 1]), 5);
+    /// ```
+    pub fn get(&self, indices: &[u64]) -> T {
+        let index = self.get_index(indices);
+        self[index].clone()
+    }
+
+    /// Get the array index from rows / columns indices.
+    ///
+    /// ```
+    /// # use tensorflow::Tensor;
+    /// # use tensorflow::eager::ReadonlyTensor;
+    /// let a: ReadonlyTensor<_> = Tensor::<f32>::new(&[3, 3, 3]).freeze();
+    ///
+    /// assert_eq!(a.get_index(&[2, 2, 2]), 26);
+    /// assert_eq!(a.get_index(&[1, 2, 2]), 17);
+    /// assert_eq!(a.get_index(&[1, 2, 0]), 15);
+    /// assert_eq!(a.get_index(&[1, 0, 1]), 10);
+    /// ```
+    pub fn get_index(&self, indices: &[u64]) -> usize {
+        assert!(self.dims.len() == indices.len());
+        let mut index = 0;
+        let mut d = 1;
+        for i in (0..indices.len()).rev() {
+            assert!(self.dims[i] > indices[i]);
+            index += indices[i] * d;
+            d *= self.dims[i];
+        }
+        index as usize
+    }
+
+    /// Returns the tensor's dimensions.
+    pub fn dims(&self) -> &[u64] {
+        &self.dims
+    }
+
+    /// Returns the tensor's dimensions as a Shape.
+    pub fn shape(&self) -> Shape {
+        Shape(Some(self.dims.iter().map(|d| Some(*d as i64)).collect()))
```

Comment:
I see you copied this from Tensor, but it can be written more succinctly nowadays as `Shape::from(&self.dims)`

Reply:
Thanks. It seems that I need to pass a slice of the dims explicitly (deref coercion did not work here).